Chore: add experimentalSkipDomainInjection option (#4955)
cypress-io/cypress-documentation: content/guides/references/experiments.md

| `experimentalRunAllSpecs`              | `false` | Enables the "Run All Specs" UI feature, allowing the execution of multiple specs sequentially.                                                                                                                                                                                                                                                                                                                                                                         |
| `experimentalOriginDependencies`       | `false` | Enables support for `require`/`import` within `cy.origin`.                                                                                                                                                                                                                                                                                                                                                                                                             |

> | `experimentalUseDefaultDocumentDomain` | `null`  | Disables injecting `document.domain` into `text/html` pages. The `document.domain` configuration has been known to cause issues with certain sites, including Google and Salesforce. Enabling this option no longer allows for default sub domain navigations, and will require the use of `cy.origin()`. This option takes an array of strings/minimatch globs. Our recommended configuration is `['*.salesforce.com', '*.force.com', '*.google.com', 'google.com']`. |

Reviewer: I would have expected more documentation for this experiment. If we know why google doesn't work and give our high-level understanding of salesforce it'll help people understand if they should enable this. I think we probably only want users to turn this on if absolutely necessary right?

Also an example of what changes might be needed where it worked without origin then needed origin might help users.

Author: > I think we probably only want users to turn this on if absolutely necessary right?

I would say so, or we leverage as a diagnostic tool. 

Do we have a place where we can further document the experiment to add some more verbosity? Right now the column already looks bloated and is a bit hard to read. I'd almost like to trim this down, and have a link to more details about the experiment that users can discover. Otherwise, we'd have to mostly guide through GitHub which I am not a big fan of.

Author: @emilyrohrbough I took a stab at this in 09e859c92f5bf80e8e6de39fa7e4544a8f57b101 and removed the reference in the web security page. I figured since this is a pretty small sample size, mentioning in the experiments and cross origin testing guide at the end should suffice

| `experimentalStudio`              | `false` | Generate and save commands directly to your test suite by interacting with your app as an end user would. |
| `experimentalRunAllSpecs`         | `false` | Enables the "Run All Specs" UI feature, allowing the execution of multiple specs sequentially.            |
| `experimentalOriginDependencies`  | `false` | Enables support for `require`/`import` within `cy.origin`.                                                |
| `experimentalSkipDomainInjection` | `null`  | Disables injecting `document.domain` into `text/html` pages.                                              |

#### Experimental Skip Domain Injection

Under the hood, Cypress
[injects document.domain](/guides/guides/web-security#Examples-of-what-Cypress-does-under-the-hood)
into your test application. In some known cases, this causes issues with certain
sites. For example, Salesforce executes code in isolated window contexts. When
Cypress injects `document.domain`, these window contexts become cross-origin and
fail to communicate. To work around this, the `experimentalSkipDomainInjection`
flag was introduced.

> This flag disables injecting `document.domain` inside Cypress, allowing Cypress
> to test these exception cases. Only enable this flag if you are testing against

Reviewer: ```suggestion
This flag disables injecting `document.domain` inside Cypress, allowing users
to test sites which don't work when `document.domain` is set. Only enable this flag if you are testing against
```

Author: addressed in 12dcee729

> Salesforce or suspect you may have `document.domain` issues. When enabled, this

Reviewer: maybe put this as a note at the bottom. do we know examples of what might be a document.domain issue?

Reviewer: Should we also say Salesforce, Google or ...?

Author: Do you mean mention them by name or have something more generic?

I don't think we have easily digestible examples of what might be a document.domain issue. Usually the site just doesn't work inside Cypress, which is _might_ be a document.domain problem

Reviewer: yeah we don't have clear examples to share. maybe removing the wording to indicate exception cases, it'll sound less like these are cases the user should know about?

```md
This `experimentalSkipDomainInjection` option disables injecting `document.domain` inside Cypress. When enabled, all cross-origin/subdomain navigation must user `cy.origin()`. This can make tests a bit more verbose and we only recommending including your site pattern if you are having issues running Cypress out of the box and suspect setting `document.domain` is interfering with your site's ability to render properly.

Before enabling, verify your application is not implementing frame busting techniques. **link**. 

At this point in time, we are aware of the following sites that require this to be set to be tested properly:
- Google
- Salesforce
```

Author: I gave this a shot in dce0200a8. Let me know what you think!

Reviewer: like it!

> option will require the use of `cy.origin()` for all cross-origin navigations,

Reviewer: ```suggestion
option will require the use of [`cy.origin()`](/api/commands/origin) for all sub-domain and cross origin navigations,
```

Author: addressed in 12dcee729

These experiments are available to be specified inside the `e2e` configuration
object:

| Option                            | Default | Description                                                                                               |
| --------------------------------- | ------- | --------------------------------------------------------------------------------------------------------- |
| `experimentalStudio`              | `false` | Generate and save commands directly to your test suite by interacting with your app as an end user would. |
| `experimentalRunAllSpecs`         | `false` | Enables the "Run All Specs" UI feature, allowing the execution of multiple specs sequentially.            |
| `experimentalOriginDependencies`  | `false` | Enables support for `require`/`import` within `cy.origin`.                                                |
| `experimentalSkipDomainInjection` | `null`  | Disables injecting `document.domain` into `text/html` pages.                                              |

#### Experimental Skip Domain Injection

Under the hood, Cypress

> [injects document.domain](/guides/guides/web-security#Examples-of-what-Cypress-does-under-the-hood)

Reviewer: I just read the Examples-of-what-Cypress-does-under-the-hood section. It doesn't state WHY we set document.domain. This is primarily to allow sub-domain navigation correct? maybe call that out here?

Author: We have a spot for this in the cross origin testing guide. I think I can move the workaround up there and it will fit fairly well

These experiments are available to be specified inside the `e2e` configuration
object:

| Option                            | Default | Description                                                                                               |
| --------------------------------- | ------- | --------------------------------------------------------------------------------------------------------- |
| `experimentalStudio`              | `false` | Generate and save commands directly to your test suite by interacting with your app as an end user would. |
| `experimentalRunAllSpecs`         | `false` | Enables the "Run All Specs" UI feature, allowing the execution of multiple specs sequentially.            |
| `experimentalOriginDependencies`  | `false` | Enables support for `require`/`import` within `cy.origin`.                                                |

> | `experimentalSkipDomainInjection` | `null`  | Disables injecting `document.domain` into `text/html` pages.                                              |

Author: addressed in b3ee9a54f.

#### Experimental Skip Domain Injection

Under the hood, Cypress
[injects document.domain](/guides/guides/web-security#Examples-of-what-Cypress-does-under-the-hood)
into your test application to lessen the burden of navigation. This is well
described in our [Cross Origin Testing](guides/guides/cross-origin-testing)
guide. However, some sites have compatibility issues with this feature.

The `experimentalSkipDomainInjection` option disables injecting
`document.domain` inside Cypress. When enabled, all cross-origin/subdomain
navigation must use `cy.origin()`, which may make tests a bit more verbose. We
only recommend including your site pattern if you are having issues running
Cypress out of the box and suspect setting `document.domain` is interfering with
your site's ability to render properly.

Before enabling, verify your application is not implementing frame busting
techniques, which you can mitigate with the

> [`experimentalModifyObstructiveThirdPartyCode`](/guides/guides/web-security#Modifying-Obstructive-Third-Party-Code)

Reviewer 2: Do we also want to link to `modifyObstructiveCode`?

Author: we probably should

Author: added in b3ee9a54f. The formatting might look a bit weird but that's `prettier` on the commit hook

flag.

> At this point in time, we are aware of the following sites that require this

Author: addressed in b3ee9a54f.

option to be set to be tested properly:

- Google
- Salesforce

> This flag can be enabled by passing an array of origin URLs or minimatch glob

Reviewer 2: Suggest linking out to `minimatch` here:

```suggestion
This flag can be enabled by passing an array of origin URLs or [minimatch](https://github.com/isaacs/minimatch) glob
```

Author: added in b3ee9a54f.
